lib/test-ast.js

```diff
@@ -0,0 +1,54 @@
+'use strict';
+
+const fs = require('fs');
+const AST = require('recast');
+
+module.exports = ({startLineNumber, title}, filePath) => {
+	const fileAst = AST.parse(tryReadFile(filePath));
+	const candidates = [];
+
+	AST.visit(fileAst, {
+		visitCallExpression(path) {
+			const {node} = path;
+
+			if (node.loc.start.line === startLineNumber) {
+				candidates.push(node);
+				return false;
+			}
+
+			this.traverse(path);
+		}
+	});
+
+	if (candidates.length === 0) {
+		throw new Error(`No test starting at line number ${startLineNumber} in ${filePath}.`);
+	}
+
+	const testAst = candidates.find(candidate => {
+		const [firstArgument] = candidate.arguments;
+		return firstArgument && firstArgument.value === title;
```

Comment:
You can declare tests without a title argument, as long as the implementation is able to provide a title for you. Or, even if there is a title argument, it may not be the exact same string. Or it may be a template literal.

In other words I don't think this will work! Why do you think it's needed? Apologies because you've probably explained this once already.

Reply:
Please re-read the description of the PR under **Approaches/considerations** for an explanation as to why.

You’re probably right about the empty case and template strings, though. So might need to rethink it – again.

Reply:
Based on the call sites we can determine the starting line of the `test()` call. Using the AST we can determine all call expressions. So for each declaration, we can determine all call expressions at that starting line. We could then figure out the end lines of each of those expressions and consider that to be the line number range for that test.

I don't think this needs the title at all?

Reply:
I realize it’s an extreme edge case, but I’m pretty sure it wouldn’t suffice for the use case I referred to in the PR description.

Like the example states: if you were to select lines 4-5 it would either run both tests or the wrong one (as it would match the one declared first).

But maybe you’re already aware and accept it or you realize something I don’t. 

Either way, your flaws are yet to be handled...

Reply:
But the AST would have the start and end lines of the first declaration (line 3), and the start and and lines of the second declaration (3 and 5), so then 4-5 would only run the second one, no?

Reply:
When you put it that way... yes, sounds perfectly reasonable. Sorry for being so slow. I'll give it a go. Thanks for bearing with me 😬

Reply:
No... I think I'm gonna have to throw in the towel on this one... I thought I got it, but trying to implement it I keep coming to the same conclusion: There's no way to distinguish between two tests which starts on the same line if that's our _only_ identifier.

I guess the part I don't get is how this would work:

>We could then figure out the end lines of each of those expressions and consider that to be the line number range for that test.

This doesn't make sense to me. How can you consider the end line of another test to be the call site test's end line?

Just to do a quick walkthrough of the example at hand:
1. We parse the test source
2. We find two candidate nodes (starting at line 3)
3. ?
4. Profit*

I need help filling in `3. ?` above. How do we go from two candidate nodes starting at the same line to determining the end line of the test in question?

Maybe you have a totally different implementation/usage of the AST in your head which is why I don't follow.

If you're convinced it would work and feel like you're talking to a wall, feel free to implement it yourself.

* Just in case: https://knowyourmeme.com/memes/profit

Reply:
Sure I can give this a go.

Reply:
Cool 😎

Just to re-phrase where my logic breaks down: I don't see how you can determine which call expression/end line number belongs to which test (if they have the same start line number).

But maybe I'm too deeply invested into this approach that I can't see the obvious...

Good luck!

Reply:
I don't see why it's relevant. There's two tests on this line? Fine, we'll run both.

Reply:
If you select the starting line (3), absolutely. But in the example I present they _start_ on the same line but don't _end_ on the same line. If you were to select a line which is part of only one of the tests, the other test shouldn't run.

But I'm gonna drop this now. I'll probably just confuse/annoy you. I sincerely hope you're right and fail to see the obvious.

Reply:
So, I had a revelation: instead of using the test title, we can pinpoint a specific test declaration by using its column number, in addition to line number.

When I look back at it, it’s so obvious with regards to the example case I posted. Since the two tests are on the same line, they must have different column numbers.

Now we can use a `Map` indexed by position (`{line, column}`) to cache line numbers. 

Thoughts?

Or are you still confident I’m lost and you can get away with only using start line number?